### ISSUE TITLE
Documentation Rework

### ISSUE THREAD

**Issue by barnabywalters** [pure-data/pure-data#1300]
Pure Data’s built-in documentation begins to fulfill the potential of interactive, “live” documentation, but is messy, inconsistent and incomplete. I’d like to rework it to fix some of these issues, but want to get an impression of whether or not the changes are likely to be accepted before putting in such a large amount of effort.

To demonstrate the sort of changes I’m talking about, here’s where I would start with the general objects:

* All:
    * make structure consistent
    * visually tidy up patches and comments
    * replace abstract data in example patches with practical real-world example usage where relevant
    * grammar fixes (e.g. capitalisation of comments)
    * thoroughly and consistently document object creation arguments, what which are required, what values are accepted, what messages are accepted etc.
- [x] `[bang]`: mention `[bng]` object, add graphical in-patch output indicator so users see a visual response from the patch without having to look in the debug console
- [x] `[float]`: mention `[f]` abbreviation
- [x] `[symbol]`: clarify “unlike float, there's no "s" message” — [float] doesn’t respond to “s” (or “f”) message, and it isn’t mentioned in the `[float]` help file, so I have no idea what this is talking about.
- [x] `[int]`: mention `[i]` abbreviation
- [ ] `[send]`: move abbreviation to top, restructure `[send]` help patch so that it’s clear that the symbol GUI element is just for debugging and isn’t required, mention $0 notation and how to use it (probably in a subpatch)
- [ ] `[receive]`: either combine with the `[send]` help patch as the two objects are almost always used together, duplicate the contents of the `[send]` patch or at least link directly to it if the `[send]` help patch remains the more comprehensive of the two
- [ ] `[select]`: in-patch visual indication of which output is active so the user doesn’t have to keep looking at the debug log
- [ ] `[route]`: ditto, more thorough explanation of what’s going on in the `[route list float symbol bang]` example
- [ ] `[pack]`: remove/clarify misleading “as with any Pd object,  you can send a list whose atoms are automatically distributed to the corresponding inlets.” — this is definitely not the case!
- [ ] `[unpack]`: clarify what that pointer arguments have something to do with data structures
- [ ] `[trigger]`: in-patch visual output
- [ ] `[spigot]`: ditto
- [ ] `[change]`: redo the example patch to better demonstrate function — as it is, it’s hard to create a situation where the same value is sent twice
- [ ] `[value]`: rename “nonlocal” to “global”, mention that scoping values to a subpatch with `$0-` doesn’t work (or is this a bug? seems like it would be useful, and would make it more consistent with send and receive, but might cause problems interfacing with `[expr]`?)

**Comment by barnabywalters** [comment]
One potential starting point would be to simply replace the existing documentation with pd-extended’s documentation files, and go through them removing any features which rely on externals, and updating the documentation for the latest version of Pure Data.

**Comment by porres** [comment]
I've been doing lots of change to the documentation, and they've been all accepted, so this is a good effort, we just need to go through these one by one

> One potential starting point would be to simply replace the existing documentation with pd-extended’s documentation files

nah... that'd be crazy, having a parallel look is fine though ;)

**Comment by porres** [comment]
> thoroughly and consistently document object creation arguments, what which are required, what values are accepted, what messages are accepted etc.

I guess the biggest debate is here, and I can tell you it's hard to actually make this "consistent" into a one-and-only pattern type. Maybe we can discuss separately, and deal with the minor details first.

**Comment by barnabywalters** [comment]
I think the way CEAMMC and pd-extended format(ted) creation argument documentation is a good template, although I agree that how creation arguments work in pd is diverse enough that there isn’t a one-size-fits-all solution. I mainly want to get a clear, consistent summary of creation arguments in one place per help patch for objects with many creation arguments/options, rather than having them scattered throughout, or mentioned in subpatches.

**Comment by porres** [comment]
> I think the way CEAMMC and pd-extended format(ted) creation argument documentation is a good template

They are different though, and I find CEAMMC's documentation very hard to read. As for Pd Extended's template, I've been using something similar to it to document objects in Cyclone and ELSE.... after doing it for over 600 help patches, I can tell you I kinda hate it! Moreover, Purr Data uses the Pd Extended format and tried revamping the documentation, but it created new issues and inconsistencies/mistakes. This is also why I said it'd be crazy to just copy Pd Extended's model and go from there as it'd be a quick and simple solution. 

All I'm saying is that documentation is something HARD, and this new model template should be more carefully thought of before hand. And sure, I agree it can improve and make some of these things more easily found.

Anyway, this is why I ask we should debate about this in a separate issue and attack the minor details you brought up here right away (like adding [f] as an abbreviation to [float], etc).

I've been working on the documentation for a while now and I also have big plans, but I'm attacking first the things that are "no-brainer", and they have all been easily incorporated into the development cycle without issues.

**Comment by barnabywalters** [comment]
Sure, having very dense, detailed help patches becomes a big maintenance hassle, especially without some degree of automation. That’s not what I was planning on starting with! I’ll try to put a PR together over the next week with the changes above so we have something concrete to discuss.

**Comment by Spacechild1** [comment]
> [pack]: remove/clarify misleading “as with any Pd object, you can send a list whose atoms are automatically distributed to the corresponding inlets.” — this is definitely not the case!

How so? It definitely *is* the case.

**Comment by porres** [comment]
> It definitely is the case.

I guess we can make it clearer that this happens when the object  does not have a list method. And maybe we can put more of such examples in other help files.

**Comment by Spacechild1** [comment]
Sorry, I didn't realize that @barnabywalters referred to the "as with any Pd object" part. This is definitely wrong, so @barnabywalters is right!

**Comment by porres** [comment]
> I find CEAMMC's documentation very hard to read

it looks good, but I don't think they did an actual good job on it.. it lacks many details... anyway, I'm just afraid we trade clarity and content for looks, which can be deceiving.

> I’ll try to put a PR together over the next week with the changes above so we have something concrete to discuss.

Can we split and have a PR with the minor fixes first? Are you planning to do it all?

Note that there is already a PR of mine that attacks many of the minor changes in separate commits, I've been doing these for a few of the last releases now.

I wouldn't mind doing many of the things you listed here, and it'd be good if we had just one single PR to merge documentation updates, as there can be lots of trouble and conflicts if you have two PRs messing with the same file. Trust me... we've been there!

**Comment by barnabywalters** [comment]
> Sorry, I didn't realize that @barnabywalters referred to the "as with any Pd object" part. This is definitely wrong, so @barnabywalters is right!

Phew, I was just about to go and test and document a variety of objects with different inlet behaviours, good think I don’t have to do that any more! To be honest I had completely forgotten that some objects distribute list atoms over their inlets — this is yet another thing which I think should definitely be mentioned in the help patches for each object which supports it.

> Can we split and have a PR with the minor fixes first? Are you planning to do it all?

My plan was to make a PR with the changes above, just for some/all of the objects in the “general” category, depending on how long it takes me. I’d see it as a “disposable” PR, small enough that I wouldn’t be sad if it gets rejected. I don’t foresee doing anything other than “minor” fixes in this PR, provided shuffling things around and doing occasional rewriting/repatching all count as “minor” :D

> Note that there is already a PR of mine that attacks many of the minor changes in separate commits, I've been doing these for a few of the last releases now.

I guess you’re talking about #1259? From what I can see, nothing there overlaps with the changes I propose above.

> I wouldn't mind doing many of the things you listed here, and it'd be good if we had just one single PR to merge documentation updates, as there can be lots of trouble and conflicts if you have two PRs messing with the same file.

You’re welcome to make the changes I suggest yourself if you’d rather I didn’t submit an additional PR, although having such a contribution bottleneck seems disadvantageous for something like documentation, where making minor fixes is accessible to many people. As a non-C programmer, documentation is one of the only things I can usefully contribute to Pd.

**Comment by porres** [comment]
> having such a contribution bottleneck seems disadvantageous for something like documentation, where making minor fixes is accessible to many people. 

minor changes/fixes in some of the files was not my concern

> Only [makefilename] is in the general category, so I can just avoid changing that file in my PR.

sorry, it is still unclear to me what you mean, what is "general category" and why are the others not there? if you're saying you just want to make minor changes in some of the files (and not all), and that you will not also change the files I already changed then it's fine.

> You’re welcome to make the changes I suggest yourself if you’d rather I didn’t submit an additional PR

yes, additional PRs for documentation can be tricky, one other thing you could do is send a PR to my documentation updates branch and we avoid conflicts and stuff

**Comment by porres** [comment]
> you could do is send a PR to my documentation updates branch

btw, it's this one yeah: https://github.com/porres/pure-data/tree/Documentation

**Comment by barnabywalters** [comment]
This is the “general category” — I’m talking about changes to the documentation, so I use the terms used in the documentation ;)

<img width="605" alt="Screenshot 2021-04-18 at 00 13 20" src="https://user-images.githubusercontent.com/968350/115127995-142bc880-9fdb-11eb-83d4-95577f9a613f.png">

> send a PR to my documentation updates branch and we avoid conflicts and stuff

Sounds like a good plan! I’ll do that.

**Comment by porres** [comment]
I see... so, you had a scope for this category first, is that it?

**Comment by barnabywalters** [comment]
> I see... so, you had a scope for this category first, is that it?

Yup, it seemed like a reasonable and achievable starting point, as well as being the first thing users see in this object list, and containing many of the most commonly used objects.

**Comment by barnabywalters** [comment]
Okay, I made a start on the first few objects. Here’s how they look. The biggest changes are to the help patch for [symbol], where I clarified the reference to the nonexistent [s< -> [float] message, and added examples and explanation of various strange behaviour.

I’d like to add a “why?” subpatch to [symbol] explaining why numbers are converted to the [symbol float<. Is this due to numbers being converted into [float x<, which [symbol] treats as a list and converts the first atom to a symbol?

<img width="675" alt="Screenshot 2021-04-18 at 14 58 12" src="https://user-images.githubusercontent.com/968350/115146590-70343280-a057-11eb-98eb-6f134dcfea24.png">
<img width="676" alt="Screenshot 2021-04-18 at 14 23 38" src="https://user-images.githubusercontent.com/968350/115146592-71655f80-a057-11eb-935c-2a9024b8bd29.png">
<img width="566" alt="Screenshot 2021-04-18 at 14 25 41" src="https://user-images.githubusercontent.com/968350/115146594-71fdf600-a057-11eb-8631-6d32c0b643a6.png">
<img width="675" alt="Screenshot 2021-04-18 at 14 38 03" src="https://user-images.githubusercontent.com/968350/115146595-732f2300-a057-11eb-93b0-06df7de675d2.png">

**Comment by porres** [comment]
Hi, I know this is a minimal layout change with a gray canvas background, but that is significant enough. I like it actually, it's clean, but like I said, we should treat design changes in a separate issue and in a later commit/PR. This is enough a difference to have to be made to **ALL** files and we're also trying to address documentation issues beyond layout. And it makes sense to apply this change at once separately once all the issues have been fixed and merged (so hold on).

Frankly, this issue is already hard enough to follow. I suggest we make separate issues and discuss each documentation issue separately, then when a comment is made it can be referred to that particular issue. It just makes things more organised and my PR follows this idea => #1259

**Comment by porres** [comment]
now for some remarks regarding [symbol]:

It's worth just mentioning it doesn't deal with floats then tell people what happens when you try to do it and why. This just adds noise. I think documentation needs to be precise and objective, right to the point. You gotta tell what it does and not waste time, space and effort saying all that it doesn't. Saying, for instance, that it cannot be abbreviated to [s] is unnecessary. If the documentation doesn't say it is abbreviated, then it's not and that's it. 

As for documenting the fact that the first element of a list is converted to a symbol, that is not true. Try sending "list a b c" and you'll see. The thing about trying to document weird behaviour is that you get into silly things like this. Again, documentation should tell you what you can do with it and people should just stick to what's documented and not abuse the object, neither the documentation has to tell you what happens when you try to abuse the object. The documentation of Pd Extended and Purr Data does that all the time and I find it terribly counterproductive.

As for converting the first element to a list, there's actually an unused code for that, see => https://github.com/pure-data/pure-data/issues/986 also note I have a Pull Request to fix that, and in the same one I'm proposing a way for symbols to convert floats to atoms. But I guess I should try to redo it. Anyway, as far as this is concerned, I think it's best to just not add noise. But, for now, I'd keep the relevant information that you can convert floats to symbols with [makefilename] but not necessarily like this.

**Comment by porres** [comment]
Well, I did lots of work in the documentation and made my best to keep connections coming from above to below and perfectly aligned. I would hope we'd keep it that way and don't make connections like these. This might be my personal opinion only, I don't know, but I do think this is visually a lot more clumsy.

I also don't understand why you need a small bang in the output, and even less the need of "(updated)" as a comment.

Anyway,  don't think there's a real significant/relevant change besides adding [f] as an abreviation. and note that when this happens in other help files, it is explicitly mention that it is a variation as an abbreviation. This is not quite clear as that here.

<img width="676" alt="115146592-71655f80-a057-11eb-935c-2a9024b8bd29" src="https://user-images.githubusercontent.com/5360142/115157269-c6f13a80-a05e-11eb-8d00-fe220f780dc1.png">

**Comment by porres** [comment]
[bang], well, we also have an abbreviation here that we missed ;) [b]

I like that we can compare to a message "bang" and we can even say that using the object is more efficient, and we could have a "see also: [bng]"

**Comment by porres** [comment]
I guess I have the same remarks of [float] for [int], but anyway, let me stress once again how I think we should start making separate issues so this doesn't become unreasonably hard to follow as a discussion.

**Comment by porres** [comment]
> Anyway, don't think there's a real significant/relevant change besides adding [f] as an abreviation.

I now see we also have the description of a default value, which is something I like. But then, my point is you need to be clear on what you're adding and changing, so we have a clear commit on what's being modified... for reference, see how I organize the commits of my PR https://github.com/pure-data/pure-data/pull/1259/commits

**Comment by porres** [comment]
sorry, my writing was terrible today, just made several significant edits to my comments above, please re read them if you already have read them

**Comment by emviveros** [comment]
@barnabywalters says
> I mainly want to get a clear, consistent summary of creation arguments in one place per help patch for objects with many creation arguments/options, rather than having them scattered throughout, or mentioned in subpatches.

Totally agree in consistent summary creation. It can be very difficult to understand when you are starting to know the object, but once you can understand it, the antiquated summary exposition is good to quick consultation.

I think the documentation shouldn't be based on just one type of user, the one which who does not yet know what the object is for, but this stage of user and the user which already understand what she/he wish and only don't remember specific flag or if that is the object you are looking for in quick view. Subpatches can do the job in that way.

**Comment by emviveros** [comment]
[@porres says](https://github.com/pure-data/pure-data/issues/1300#issuecomment-822043504)
> I also don't understand why you need a small bang in the output, and even less the need of "(updated)" as a comment.
If you update with the same number you have a visual update feedback. It's useful to debug.

**Comment by porres** [comment]
hmmm, so "updated" is to say that there was an update on the output... well, that seems redundant, but anyway, it wasn't clear, I thought it'd mean that that part of the help file had been updated :)  

but yeah, I have to admit I knew the purpose of the bang, but it's just that most of the time it'll flash also in redudancy. I did check to see if that was comming from Purr Data or Extended, but that wasn't even a concern there. I have to say a print object seems better here. I know there seems to be an idea here that looking at the console is a lot of hassle, but c'mon, this is Pd, and similar environments have a console window that we, as users and developers have to look at all the time, it's a good practice.

I see that sometimes that is a drag as it'd be just easier if we could see what's coming out, just like we have in the example of float a number box to show us, but in the example of [route] we can't do that mostly because Pd can only show us symbols and floats, it lacks a bultin GUI that shows lists, for instance. But putting bangs everywhere doesn't really fix this, you'd still have to look at the terminal for a proper read out.

**Comment by emviveros** [comment]
> I see that sometimes that is a drag as it'd be just easier if we could see what's coming out, just like we have in the example of float a number box to show us, but in the example of [route] we can't do that mostly because Pd can only show us symbols and floats, it lacks a bultin GUI that shows lists, for instance.

Totally agreed!

**Comment by barnabywalters** [comment]
Wow, I was not expecting so much feedback! It’s difficult to parse and respond to when split up into so many replies, but I will try.

In my opinion, improving the help patches by taking multiple passes over them (making whatever fits into your as-yet unclear category of “minor fixes” before doing whatever counts as a “significant change”) is inefficient. If I have a help patch open to correct the content, it takes very little additional time and effort to make layout improvements such as those I suggested above (consistent visual structure, balanced white space, aligning comments with objects, etc.), and due to Pd’s visual nature, the two are quite closely linked. If you insist on fixing content and structure separately then unfortunately I won’t be contributing as I don’t see it as an efficient use of my limited time.

That’s also why I haven’t yet made separate issues for problems I found in every single patch: the layout and structural improvements I am interested in making are broadly generalisable, and require a consensus before per-patch discussion about content issues.

Regarding your feedback on [symbol]:

I agree about not taking up space describing what the object does not do. I was correcting mistakes in the pre-existing paragraph which talks about [symbol] not responding to the [send< message, and thought “if we’re talking about differences between [symbol] and [float], [int] etc. then let’s mention the lack of an abbreviation too”, but agree that in fact neither paragraph is necessary.

Regarding conversion of the first item from a list, that’s my mistake, I forgot that list messages are only true lists if they start with the list atom. What are messages like [A B C< officially referred to as? Multi-atom messages?

> Again, documentation should tell you what you can do with it and people should just stick to what's documented and not abuse the object.

Pd is a very permissive environment, with many quirks. In cases where objects silently behave in unexpected ways (such as [symbol] converting an incoming float into the "float" symbol), I think that users benefit from the documentation clarifying what is happening. When an object behaves in a way I don’t expect, my first recourse is to open up the help patch; if it doesn’t even acknowledge the unexpected behaviour, then it looks like a bug. In this specific case, I agree that the [makefilename] example is sufficient to describe how [symbol] interacts with floats, and that the [6< message should be removed, or moved to a subpatch which talks in more detail about how [symbol]’s conversion process works.

If Pd had an official reference manual, with exhaustive long-form documentation of each object, then I’d agree that help patches would be better off as quick reference guides. This is not the case, so I’d argue that for the moment, it’s worth taking up some extra space in help patches (or their subpatches) to document unexpected behaviour.

Either way, it’s good to hear that a lot of [symbol]’s unexpected behaviour will hopefully be fixed soon!

Regarding your feedback on [float]:

I agree that the diagonal connection is ugly and best avoided. The reason I used it in this case was because the [value] demonstration is so tangential to the help patch that I didn’t want to dedicate an entire extra line of vertical space to it. Looking back at it now, I think that the [receive] example alone is sufficient and preferable for showing how [float] reacts to a [send label< message, and the [value] example can just be removed from this and other related patches.

Regarding the small [bng] I added to the output: I have no idea why you mentioned looking in extended or purr-data to see “if it was from there” — what does that have to do with anything?! My reasoning for suggesting it is that [float] (and [int]) don’t entirely follow the “message to first inlet triggers an output” rule-of-thumb which is otherwise broadly true for pd’s objects, as [send label< to the first inlet doesn’t trigger an output. Adding a small, unobtrusive [bng] to the output takes only a single object and ~25px of space, and provides unambiguous contextual visual feedback about how [float] behaves — what inputs trigger (or, just as importantly, *don’t* trigger) an output. This is **not** otherwise the case, as the number GUI control doesn’t visually indicate when the same number is sent to it twice (as @emviveros already pointed out).

> I have to say a print object seems better here. I know there seems to be an idea here that looking at the console is a lot of hassle, but c'mon, this is Pd, and similar environments have a console window that we, as users and developers have to look at all the time, it's a good practice.

Take a moment to practically consider what you’re suggesting: that the tiny [bng] I added should be replaced with a much larger [print] object, which *no longer provides contextual live visual feedback* (i.e. it’s just visual noise in the patch) and *requires the user to switch windows, or to have an unfocused window visible in the background to get *any* benefit out of it* — and then, when they do switch over to the console window, they have to figure out whether or not a new “5” was added to the long list of “5”s and other numbers already clogging up the console from their experimentation with the help patch! I see only disadvantages to this. There are definitely occasions where using the [print] object in a help patch is the best way to show how an object is behaving. I’m confident that [float] is not one of them.

[philosophical tangent]

Pd’s visual nature can be a huge learning advantage over programming environments where looking at a console is the only way of getting feedback. It’s an opportunity to write live, interactive documentation which allows the user to gain an intuitive understanding of how objects work based on visual pattern-matching abilities which have been evolving for millennia. Saying that forcing a user to switch contexts to a console window to get feedback from the documentation they’re playing with is a “good practice”, when there’s a better approach with no additional development or maintenance cost, based on no actual evidence or reasoning other than that it “seems” better to you… I’m not sure how to respond to that, other than being sad about a missed opportunity.

[/philosophical tangent]

>  But then, my point is you need to be clear on what you're adding and changing, so we have a clear commit on what's being modified..

These screenshots were intended as a quick demonstration of the approach I’d like to take to improving the documentation, so I didn’t write an exhaustive list of every tiny change I made (it’d be a looong list of “aligned comment with object, fixed grammar, aligned comment with object, made margins consistent, aligned comment with object, etc.). In the event that I end up making actual commits, I’m happy to summarise my changes in the commit, and provide before/after screenshots to make reviewing changes easier. I totally understand how awkward reviewing changes to pd patches as plaintext is, and how that complicates the process of reviewing documentation changes, and am happy to make the process easier.

-----

As for how to proceed, my suggestion would be as follows:

1. Reach a rough consensus, either here or in another issue, about help patch layout and visual style.
2. Make a template help patch to serve as a style guide, with guidelines for structure, margins and white space for contributors to refer to.
2. Make an issue for each individual help patch, where issues related to content can be discussed
3. Make a PR for each patch, fixing content issues and layout simultaneously
4. Merge PRs to a documentation branch, preferably here if possible as I’d rather not send PRs to a fork and split off discussion and governance.
5. Merge documentation changes to master whenever the maintainers feel it makes sense

Just for clarity: who here is able to merge pull requests? I don’t see a list of maintainers or a contribution guide, and would like to make sure that I’m getting some feedback from someone who’s able to merge PRs.

**Comment by porres** [comment]
> In my opinion, improving the help patches by taking multiple passes over them (making whatever fits into your as-yet unclear category of “minor fixes” before doing whatever counts as a “significant change”) is inefficient. 

I'm not saying it isn't, but it's a practice that helps in getting it done and merged. It's a hint. It may be efficient, but also counterproductive.

"minor fixes": silly but important things like "add a [f] abbreviation", fix typos, fix a wrong information/issue, document an undocumented feature. In short, regular documentation issues concerning content/information. Even more objective things like this can be open to debate and agreement. For reference, see my proposed change to route from yesterday that's still going on => https://github.com/pure-data/pure-data/issues/1302 
“significant change”: a whole structural design change in all the patches at once, addressing not content issues, but format/design principles. I say we hold this for later, as that surely opens to a broader consensus discussion.

> If you insist on fixing content and structure separately then unfortunately I won’t be contributing as I don’t see it as an efficient use of my limited time.

Yeah, that's exactly what I'm proposing. But that's just me and my 2 cents. And is not like it's up to me, it's what I think it works best from experience by contributing myself to this project. See, if you fix many things that are "no brainer" but also have many personal design choices, the commit that could be easily accepted to fix the content problems can be rejected for the other issues. Hence, separating every step makes it better for acceptance than a huge single pack with a gazillion changes.

> Just for clarity: who here is able to merge pull requests? I don’t see a list of maintainers or a contribution guide, and would like to make sure that I’m getting some feedback from someone who’s able to merge PRs.

Miller Puckette is the author and the only one who can decide and give the last call on what gets merged or not. The PRs that have better chances, from my experience, follows the idea I'm proposing here. I could give annedotical examples.

**Comment by porres** [comment]
> Pd is a very permissive environment, with many quirks. In cases where objects silently behave in unexpected ways (such as [symbol] converting an incoming float into the "float" symbol), I think that users benefit from the documentation clarifying what is happening. When an object behaves in a way I don’t expect, my first recourse is to open up the help patch; if it doesn’t even acknowledge the unexpected behaviour, then it looks like a bug. 

If the documentation doesn't say deals with float and just say it stores symbols, you should not assume that 1) it handles floats 2) it has a bug in doing so.

We're not talking about an edge case or a "silent operation", we're talking about something that is simply not supported and, consequently, not documented. Objects can behave in many weird way when handling unsupported data and it's not the job of the documentation to tell you what happens when you try to do things you shouldn't do. It should just be clear on what it does. If the fact that [symbol] only deals with symbols is still ambiguous, we can try and improve that. 


> In this specific case, I agree that the [makefilename] example is sufficient to describe how [symbol] interacts with floats, and that the [6< message should be removed, or moved to a subpatch which talks in more detail about how [symbol]’s conversion process works.

There's no conversion process, symbols do not convert floats to symbols, it never did. Hopefully it will if my PR gets merged.

 
> If Pd had an official reference manual, with exhaustive long-form documentation of each object, then I’d agree that help patches would be better off as quick reference guides. This is not the case, so I’d argue that for the moment, it’s worth taking up some extra space in help patches (or their subpatches) to document unexpected behaviour.

Again, feels more like documented unsupported actions. It is not supposed to be expected that [symbol] handles floats. It is not acting in any unexpected way. since it never promised to deal with them with a example showing that it works. You can check the source code, it doesn't even have a float method https://github.com/pure-data/pure-data/blob/master/src/x_connective.c#L177

Note that floats aren't rejected to give you a typical warning like "[symbol] expected symbol but got float" because the object takes anythings (so it converts a single symbolic atom to a symbol as the documentation shows). Now, I don't know, maybe in this context we could propose a better safeguard and a float method just to filter it out and give a warning, but that's something else. In the documentation, it is sufficient to say it doesn't handle floats, which is implicit by the simple lack of such an example, but could be more explicitly stated.

**Comment by porres** [comment]
well, been doing somre of the stuff suggested here on my documentation update branch...

**Comment by porres** [comment]
> well, been doing somre of the stuff suggested here on my documentation update branch...

just did what was just simpler, more 'crucial' mistakes and no brainers. Remaining issues and debates can furtherly be discussed her, but anyway, I'd hope we could have separate issues for each thing.

**Comment by emviveros** [comment]
Particularly I think the @barnabywalters idea of start to discuss a unified new model of documentation can be helpful. It doesn't have to be done in a hurry and updating little things in documentation as Porres has been doing is necessary to make Pd better for new users. But despite @porres' efforts, we still need better documentation.

**Comment by porres** [comment]
> Regarding the small [bng] I added to the output: I have no idea why you mentioned looking in extended or purr-data to see “if it was from there” — what does that have to do with anything?! 

Cause it seemed like something they'd do and I actually had it in my mind that I've seen this once somewhere sometime ago and thought it was silly and it really did bug me. Now I know where I had seen this: MAX's documentation. It has this bang that flashes every time along with the number box. But ok, you didn't take this from anywhere.

> My reasoning for suggesting it is that [float] (and [int]) don’t entirely follow the “message to first inlet triggers an output” rule-of-thumb which is otherwise broadly true for pd’s objects

They actually do though.

> as [send label< to the first inlet doesn’t trigger an output.

 If you're saying that not every message to the left inlet triggers an output, that is also quite common for other objects. The messages in the documentation tells you what happens in all cases though, and it clearly states in [float], for instance, that incoming floats does "set and output value", that a bang message does "output the value". It also says that 'send' does send the value carried in [f] somewhere else and it doesn't say the value is sent out the object again. If it isn't clear yet, we can more explicitly add the "there's no output". 

> Adding a small, unobtrusive [bng] to the output takes only a single object and ~25px of space, and provides unambiguous contextual visual feedback about how [float] behaves 

I really think it's best to make things clear in the text if the problem is lack of clarity or ambiguity in the text. Unless people want too call help files to not look at the text and try to figure it out by not reading them...

> Take a moment to practically consider what you’re suggesting: that the tiny [bng] I added should be replaced with a much larger [print] object, which no longer provides contextual live visual feedback

I guess I wasn’t all too clear, I was speaking more in general, and gave the example of [route] to make my point. Regarding [float], it just seems unnecessary, both the bang and the print.

**Comment by porres** [comment]
> I guess I wasn’t all too clear, I was speaking more in general, and gave the example of [route] to make my point. Regarding [float], it just seems unnecessary, both the bang and the print.

And it's not like I'm against it in any case, in my update of the [bang] object in one my last commits I did put the bang and the [print] object...

**Comment by porres** [comment]
@emviveros I agree that bringing this to the Pd list can be better to have more input.

But I have to say I'm kinda lost here as to what we are talking about. It's been so much and issues that are very case specific. Like debating wether a float method to a symbol is a bug, an unexpected behaviour and wether some information on the help patch is clear or not... 

So I'm lost now.

I guess we'd have to start over and make it more objective.

**Comment by umlaeute** [comment]
>     1. start by answering what is the profile of the Pd user and what he needs within the documentation.

you might start by not assuming that "the Pd user" is a white male academic.

**Comment by emviveros** [comment]
[@porres says:](https://github.com/pure-data/pure-data/issues/1300#issuecomment-846441019)
> I guess we'd have to start over and make it more objective.

Totally agree! Maybe start a new issue with title like Pattern design for Pd Documentation 

@umlaeute says:
> you might start by not assuming that "the Pd user" is a white male academic.

hahahaha!! Yes! Please!!! Maybe we have to extend the answering to Pd Facebook Groups. Maybe there we can expand the scope of the research. Looking for documentation of softwares created by people who do not understand themselves in the scope of white academic men

**Comment by emviveros** [comment]
I opened the [new issue](https://github.com/pure-data/pure-data/issues/1320)  to centralize discussion as suggested by @porres . Please, contributions starting by possible topic rename are welcomed!

**Comment by porres** [comment]
I'm there, maybe we can close this then?

**Comment by porres** [comment]
again, can't we just close this? ping @umlaeute

**Comment by porres** [comment]
please @umlaeute / @Spacechild1 let's just close this. Most of this is being taken care of on 'pddp' and/or has already been fixed and merged into master!

**Comment by porres** [comment]
"most" meaning "all that is relevant", because there's just lots of parallel discussions...

**Comment by Spacechild1** [comment]
Ok, closing this. The issue tracker isn't the right place for such broad discussion, anyway. We have GitHub "Discussions" for that :-)
